language/stdlib/modules/authenticator.move

```diff
@@ -0,0 +1,69 @@
+// Move version of the Libra Transaction authenticator logic. Supports:
+// - creating an authentication policy `T` from public keys a a threshold
+// - converting an authentication policy to an account authentication key
+// - allowing one of the parties in a multisig to rotate their own key
+module Authenticator {
+    use 0x0::Hash;
+    use 0x0::Signature;
+    use 0x0::Transaction;
+    use 0x0::Vector;
+
+    // An authentication policy
+    struct T {
+        public_keys: vector<vector<u8>>,
+        threshold: u8,
+    }
+
+    public fun create(public_keys: vector<vector<u8>>, threshold: u8): T {
+        T { public_keys, threshold }
+    }
+
+    // Convert the policy `t` into an account authentication key for scheme `scheme_id`
+    fun authentication_key(t: &T, scheme_id: u8): vector<u8> {
+        let keys = &t.public_keys;
+        let len = Vector::length(keys);
+        let authentication_key_preimage = Vector::empty();
+        let i = 0;
+        while (i < len) {
+            Vector::append(
+                &mut authentication_key_preimage,
+                *Vector::borrow(keys, i)
+            );
+            i = i + 1;
+        };
+        Vector::push_back(&mut authentication_key_preimage, scheme_id);
+        Hash::sha3_256(authentication_key_preimage)
+    }
+
+    // Convert the policy `t` into a multisig account authentication key
+    public fun multi_ed25519_authentication_key(t: &T): vector<u8> {
+        // 1u8 is the scheme id for multi_ed25519
+        authentication_key(t, 1u8)
+    }
+
+    // Convert the policy `t` into a single-sig account authentication key
+    public fun ed25519_authentication_key(t: &T): vector<u8> {
+        Transaction::assert(t.threshold == 1, 7001);
+        Transaction::assert(Vector::length(&t.public_keys) == 1, 7002);
+        // 0u8 is the scheme id for ed25519
+        authentication_key(t, 0u8)
+    }
+
+    // Allow the `public_key_idx`th public key of a multisig authenticator to rotate their key to
+    // `new_public_key` using a correct signature from the old key.
+    // `signature` should be the signature of the `public_key_idx`th's corresponding private key on
+    // `message`
+    public fun rotate_multi_ed25519(
```

Comment:
We should reconsider this new feature due to major business and security related side effects. Enlisting a few of them:

1. It shouldn't be possible to arbitrarily increase the `seq_id` of an account if you don't meet the threshold. There might be an important business reason for that (i.e., pending atomic swaps, digital remittances or other smart contract logic, pre-signed transactions etc).
2. It shouldn't be possible to arbitrarily alter the `auth_key` as there might be inflight or planned transactions (cheques) that would be affected by such a change. A minimum of key owners who meet the threshold should always be aware on any state change of their account.
3. One party can drain the account by spending with unlimited gas using subsequent uncontrolled key rotations.

1 and 3 can be solved by only allowing single key rotation via another account, but there is no obvious solution for the 2nd issue.

Reply:
Good points on all three!
- Agreed that (2) is a huge issue in the general case. If one of the keys gets compromised or is owned by someone you don't trust, they can cause conflicts that make it very hard to get a multisigned transaction through
- For the particular use-case that this is trying to address (on-chain governance by Association members sharing a multisig account), I think this approach could be reasonable. In this setting, griefing has social consequences, and we could take some precautions to guard against the compromise take. One measure we could take is to rate-limit rotation of a particular key (e.g., at most once a week) to make griefing with a compromised key ineffective (you get one chance, then everyone will notice and the policy will be updated  to remove you). But definitely worth discussing more.
- If we need the flexibility for individuals in a group policy to update their keys, but can't tolerate the auth_key conflicts, we can always consider on-chain, multi-tx multisig instead.
- I didn't finish implementing this, but it can be done in a way that rules out (1) and (3) by construction using your suggested approach (single key rotation from different account only)[](url)

Reply:
We could imagine a case wherein an account can opt into this. Such that they take the risks associated with supporting this.

Reply:
Is it possible to drain an account even with one key rotation? Ie by maliciously setting a huge gas amount?

Reply:
> We could imagine a case wherein an account can opt into this. Such that they take the risks associated with supporting this.

Agreed that opt-in makes sense! It is actually challenging to implement this in a way that  is *not* opt-in (we'd have to change `LibraAccount`)

> Is it possible to drain an account even with one key rotation? Ie by maliciously setting a huge gas amount?

It is possible. However, keep in mind the account being drained is *not* the collectively owned multisig account, it is the account owned by the person whose individual public key is being rotated. So they can only waste their own funds by doing this.